napari/utils/io.py

```diff
@@ -291,13 +292,54 @@ def write_csv(
             writer.writerow(row)
 
 
+def csv_to_layer_data(filename: str) -> List[FullLayerData]:
+    """Return layer data from a csv.
+
+    Parameters
+    ----------
+    filename : str
+        Filename for reading csv.
+
+    Returns
+    -------
+    list of napari.types.FullLayerData or None
+        Layer full layer data or None.
+    """
+
+    gen = iter_csv(filename)
+    column_names = next(gen)
```

Comment:
Although I initially liked the generator approach, I'm concerned that this might leave the file open until Python quits. I don't know whether Python will be smart enough to garbage collect the generator when it falls out of scope.

See [here](http://scipy-lectures.org/advanced/advanced_python/index.html#bidirectional-communication) that generators have a `.close()` method, so it might be worth adding:

```suggestion
    gen = iter_csv(filename)
    column_names = next(gen)
    gen.close()
```

Reply:
For completeness' sake, here's the walrus equivalent, which we can't use because you need Py3.8, but it's still nice to note the utility of it :smiley: :

```python
column_names = next(gen := iter_csv(filename))
gen.close()
```

It might also be worth making `iter_csv` itself a context manager...

Reply:
> See [here](http://scipy-lectures.org/advanced/advanced_python/index.html#bidirectional-communication) that generators have a `.close()` method, so it might be worth adding:

funny, I had that in there at first and removed it.  Should have stuck with my gut :)

Another thing that I wanted to mention is that maybe we should use this little `yield from` wrapper on csv.reader at all?  Maybe we should just use the `with open()... reader = csv.reader` each time?  Makes it clearer that this is just Stalin

Reply:
I've added the `gen.close`, but it does seem like the with open syntax here would be more appropriate - @tlambert03 could you maybe do that swap?

Reply:
updated